### PR TITLE
simplify sender interface

### PR DESCRIPTION
**Pull request by k15r**
**Description**

Changes proposed in this pull request:

previously the sender had to return
- a PublishResult
- an Error

The problem here was that PublishResult was treated as an error
indicator for the eventMesh Backend, and the error was for other senders
such as the on Nats

The new interface just expects an Error to be returned in case of an
error. Any successful call will result in a 204 response with an empty
message.

This means now the sender has to map its errors or error responses onto
an error object, and any other function in the chain can just treat all
errors equally.

**Related issue(s)**

### CODE REVIEW

**Review by mmitoraj — APPROVED** [review]

File: components/event-publisher-proxy/pkg/sender/eventmesh/eventmesh.go [hunk 1]
```diff
@@ -36,47 +37,43 @@ func (s *Sender) URL() string {
...
+		return e
+	}
+	if resp.StatusCode >= http.StatusOK && resp.StatusCode < http.StatusMultipleChoices {
+		return nil
 	}
 	body, err := io.ReadAll(resp.Body)
 	defer func() { _ = resp.Body.Close() }()
 	if err != nil {
-		return nil, err
+		return common.ErrInternalBackendError
```
Comment (mfaizanse) [inline components/event-publisher-proxy/pkg/sender/eventmesh/eventmesh.go:74]: should we log the `err` object here? or wrap it with the returned error?

File: components/event-publisher-proxy/pkg/sender/jetstream/jetstream.go [hunk 2]
```diff
@@ -58,48 +60,60 @@ func NewSender(ctx context.Context, connection *nats.Conn, envCfg *env.NATSConfi
...
+func (s *Sender) Send(_ context.Context, event *event.Event) sender.PublishError {
 	if s.ConnectionStatus() != nats.CONNECTED {
-		return nil, ErrNotConnected
+		return ErrNotConnected
 	}
 
 	jsCtx, jsError := s.connection.JetStream()
 	if jsError != nil {
-		return nil, jsError
+		return common.ErrClientNoConnection
```
Comment (mfaizanse) [inline components/event-publisher-proxy/pkg/sender/jetstream/jetstream.go:77]: should we log the jsError object here? or wrap it with the returned error?

File: components/event-publisher-proxy/pkg/handler/handler.go [hunk 3]
```diff
@@ -331,10 +330,9 @@ func (h *Handler) sendEventAndRecordMetrics(ctx context.Context, event *cev2even
 			originalEventType = event.Type()
 		}
 	}
-	h.collector.RecordEventType(originalEventType, event.Source(), result.HTTPStatus())
-	h.collector.RecordBackendLatency(duration, result.HTTPStatus(), host)
-	h.collector.RecordBackendRequests(result.HTTPStatus(), host)
```
Comment (mfaizanse) [inline components/event-publisher-proxy/pkg/handler/handler.go:336]: Dont we need this metric `h.collector.RecordBackendRequests(result.HTTPStatus(), host)`?

like `h.collector.RecordBackendRequests(http.StatusNoContent, host)`?

File: components/event-publisher-proxy/pkg/metrics/collector.go [hunk 4]
```diff
@@ -13,21 +13,11 @@ import (
...
 	// BackendLatencyKey name of the backendLatencyHelp metric.
 	BackendLatencyKey = "eventing_epp_backend_duration_milliseconds"
 	// backendLatencyHelp help text for the backendLatencyHelp metric.
 	backendLatencyHelp = "The duration of sending events to the messaging server in milliseconds"
 
-	// BackendRequestsKey name of the eventRequests metric.
-	BackendRequestsKey = "eventing_epp_backend_requests_total"
-	// backendRequestsHelp help text for event backendRequests metric.
-	backendRequestsHelp = "The total number of backend requests"
-
```
Comment (mfaizanse) [inline components/event-publisher-proxy/pkg/metrics/collector.go:30]: We dont need these metrics? or do you think the info is already covered by the latency metric?

File: components/event-publisher-proxy/pkg/sender/eventmesh/eventmesh.go (hunk 1 above)
Comment (k15r) [reply components/event-publisher-proxy/pkg/sender/eventmesh/eventmesh.go:74]: yeah, logging is probably a good idea

File: components/event-publisher-proxy/pkg/sender/jetstream/jetstream.go (hunk 2 above)
Comment (k15r) [reply components/event-publisher-proxy/pkg/sender/jetstream/jetstream.go:77]: yeah, logging is probably a good idea

File: components/event-publisher-proxy/pkg/metrics/collector.go (hunk 4 above)
Comment (k15r) [reply components/event-publisher-proxy/pkg/metrics/collector.go:30]: the latency metric is a histogram.. it already provides you a count 
so figuring out the failed ones is just about filtering by the statusCode

File: components/event-publisher-proxy/pkg/handler/handler.go (hunk 3 above)
Comment (k15r) [reply components/event-publisher-proxy/pkg/handler/handler.go:336]: BackendRequests is the total count of requests done to the backend -> 
eventing_epp_backend_duration_milliseconds_count

as the latency metric has destination_service and code as labels we can filter by those and get:
- all successful requests (filter by 2XX)
- all failed requests (filter by non2XX)
- all requests (dont filter)

File: components/event-publisher-proxy/pkg/sender/jetstream/jetstream.go (hunk 2 above)
Comment (mfaizanse) [reply components/event-publisher-proxy/pkg/sender/jetstream/jetstream.go:77]: Do we log here now?

Comment (k15r) [reply components/event-publisher-proxy/pkg/sender/jetstream/jetstream.go:78]: yes :)

**Review by NHingerl — APPROVED** [review]

**Review by mfaizanse — APPROVED** [review]

**Pull request merged**